File: backend/app/domain/validation/pattern.py
"""Scanner and compiler for the bounded Python/JavaScript pattern subset."""

import re
from dataclasses import dataclass, field

from app.domain.errors import RuleViolationError

MAX_PATTERN_LENGTH = 256
MAX_ALTERNATIVES = 16
MAX_ATOMS_PER_ALTERNATIVE = 64
MAX_QUANTIFIED_ATOMS_PER_ALTERNATIVE = 16
MAX_REPETITION = 256
MAX_BRANCH_MATCH_LENGTH = 1_024

_ESCAPABLE_METACHARACTERS = frozenset(r".\^$*+?{}[]|()-")
_FORBIDDEN_UNESCAPED = frozenset("^$*+?{}()]")


@dataclass(frozen=True, slots=True)
class PortablePattern:
    source: str
    max_match_length: int
    _compiled: re.Pattern[str] = field(repr=False, compare=False)

    def fullmatch(self, value: str) -> bool:
        return len(value) <= self.max_match_length and self._compiled.fullmatch(value) is not None


@dataclass(frozen=True, slots=True)
class _AcceptedCharacters:
    intervals: tuple[tuple[int, int], ...]
    matches_any: bool = False

    def overlaps(self, other: "_AcceptedCharacters") -> bool:
        if self.matches_any or other.matches_any:
            return True
        return any(
            left_start <= right_end and right_start <= left_end
            for left_start, left_end in self.intervals
            for right_start, right_end in other.intervals
        )


_ANY_CHARACTER = _AcceptedCharacters((), matches_any=True)


def _invalid(diagnostic: str) -> RuleViolationError:
    return RuleViolationError(diagnostic, code="portable_pattern_invalid")


class _PatternScanner:
    def __init__(self, source: str) -> None:
        self.source = source
        self.index = 0

    def scan(self) -> int:
        branch_maxima: list[int] = []
        while True:
            branch_maxima.append(self._scan_branch())
            if self.index == len(self.source):
                break
            self.index += 1  # top-level alternation separator
            if len(branch_maxima) >= MAX_ALTERNATIVES:
                raise _invalid("portable pattern has too many alternatives")
        return max(branch_maxima)

    def _scan_branch(self) -> int:
        atom_count = 0
        quantified_count = 0
        maximum_length = 0
        previous_quantified = False
        previous_characters: _AcceptedCharacters | None = None

        while self.index < len(self.source) and self.source[self.index] != "|":
            characters = self._scan_atom()
            atom_count += 1
            if atom_count > MAX_ATOMS_PER_ALTERNATIVE:
                raise _invalid("portable pattern branch has too many atoms")

            maximum_repetition, quantified = self._scan_quantifier()
            if quantified:
                quantified_count += 1
                if quantified_count > MAX_QUANTIFIED_ATOMS_PER_ALTERNATIVE:
                    raise _invalid("portable pattern branch has too many quantified atoms")
                if (
                    previous_quantified
                    and previous_characters is not None
                    and previous_characters.overlaps(characters)
                ):
                    raise _invalid("adjacent quantified atoms accept overlapping characters")

            maximum_length += maximum_repetition
            if maximum_length > MAX_BRANCH_MATCH_LENGTH:
                raise _invalid("portable pattern branch maximum length is too large")
            previous_quantified = quantified
            previous_characters = characters

        if atom_count == 0:
            raise _invalid("portable pattern alternatives cannot be empty")
        return maximum_length

    def _scan_atom(self) -> _AcceptedCharacters:
        char = self.source[self.index]
        if char == ".":
            self.index += 1
            return _ANY_CHARACTER
        if char == "[":
            return self._scan_character_class()
        if char == "\\":
            literal = self._scan_escape()
            return _AcceptedCharacters(((ord(literal), ord(literal)),))
        if char in _FORBIDDEN_UNESCAPED:
            raise _invalid(f"unescaped metacharacter {char!r} is not portable")

        self.index += 1
        return _AcceptedCharacters(((ord(char), ord(char)),))

    def _scan_escape(self) -> str:
        self.index += 1
        if self.index == len(self.source):
            raise _invalid("portable pattern ends with an incomplete escape")
        escaped = self.source[self.index]
        if escaped not in _ESCAPABLE_METACHARACTERS:
            raise _invalid("only literal metacharacters may be escaped")
        self.index += 1
        return escaped

    def _scan_character_class(self) -> _AcceptedCharacters:
        self.index += 1
        items: list[tuple[str, bool]] = []
        while self.index < len(self.source) and self.source[self.index] != "]":
            char = self.source[self.index]
            if char == "\\":
                items.append((self._scan_escape(), True))
                continue
            if char == "[" or (char == "^" and not items):
                raise _invalid("nested or negated character classes are not portable")
            items.append((char, False))
            self.index += 1

        if self.index == len(self.source):
            raise _invalid("portable character class is not closed")
        self.index += 1
        if not items:
            raise _invalid("portable character classes cannot be empty")

        intervals: list[tuple[int, int]] = []
        item_index = 0
        while item_index < len(items):
            char, escaped = items[item_index]
            if (
                item_index + 2 < len(items)
                and items[item_index + 1] == ("-", False)
                and char != "-"
            ):
                end, _ = items[item_index + 2]
                if end == "-" or ord(char) > ord(end):
                    raise _invalid("portable character class contains a descending range")
                intervals.append((ord(char), ord(end)))
                item_index += 3
                continue
            if char == "-" and not escaped and item_index not in {0, len(items) - 1}:
                raise _invalid("portable character class contains a malformed range")
            intervals.append((ord(char), ord(char)))
            item_index += 1

        return _AcceptedCharacters(tuple(intervals))

    def _scan_quantifier(self) -> tuple[int, bool]:
        if self.index == len(self.source):
            return 1, False
        if self.source[self.index] == "?":
            self.index += 1
            return 1, True
        if self.source[self.index] != "{":
            return 1, False

        closing = self.source.find("}", self.index + 1)
        if closing == -1:
            raise _invalid("portable repetition is not closed")
        body = self.source[self.index + 1 : closing]
        parts = body.split(",")
        if len(parts) == 1 and parts[0].isascii() and parts[0].isdigit():
            lower = upper = int(parts[0])
        elif (
            len(parts) == 2
            and parts[0].isascii()
            and parts[0].isdigit()
            and parts[1].isascii()
            and parts[1].isdigit()
        ):
            lower, upper = (int(part) for part in parts)
        else:
            raise _invalid("portable repetition must be {m} or {m,n}")
        if lower > upper:
            raise _invalid("portable repetition bounds are descending")
        if upper > MAX_REPETITION:
            raise _invalid("portable repetition upper bound is too large")
        self.index = closing + 1
        return upper, True


def compile_portable_pattern(source: str) -> PortablePattern:
    """Validate a bounded pattern before entrusting it to the host regex engine."""
    if not source:
        raise _invalid("portable pattern cannot be empty")
    if len(source) > MAX_PATTERN_LENGTH:
        raise _invalid("portable pattern source is too long")

    max_match_length = _PatternScanner(source).scan()
    try:
        compiled = re.compile(source)
    except re.error as exc:  # scanner diagnostics should normally win
        raise _invalid(f"portable pattern could not be compiled: {exc.msg}") from exc
    return PortablePattern(
        source=source,
        max_match_length=max_match_length,
        _compiled=compiled,
    )

File: frontend/src/shared/domain/validation/pattern.ts
const MAX_PATTERN_LENGTH = 256
const MAX_ALTERNATIVES = 16
const MAX_ATOMS_PER_ALTERNATIVE = 64
const MAX_QUANTIFIED_ATOMS_PER_ALTERNATIVE = 16
const MAX_REPETITION = 256
const MAX_BRANCH_MATCH_LENGTH = 1_024

const ESCAPABLE_METACHARACTERS = new Set([...String.raw`.\^$*+?{}[]|()-`])
const FORBIDDEN_UNESCAPED = new Set([...String.raw`^$*+?{}()]`])

interface AcceptedCharacters {
  readonly intervals: readonly (readonly [number, number])[]
  readonly matchesAny: boolean
}

const ANY_CHARACTER: AcceptedCharacters = { intervals: [], matchesAny: true }

export interface PortablePattern {
  readonly source: string
  readonly maxMatchLength: number
  matches(value: string): boolean
}

export class PortablePatternError extends Error {
  readonly code = 'portable_pattern_invalid'

  constructor(diagnostic: string) {
    super(diagnostic)
    this.name = 'PortablePatternError'
  }
}

class PatternScanner {
  private readonly characters: readonly string[]
  private index = 0

  constructor(source: string) {
    this.characters = [...source]
  }

  scan(): number {
    const branchMaxima: number[] = []
    while (true) {
      branchMaxima.push(this.scanBranch())
      if (this.index === this.characters.length) break
      this.index += 1
      if (branchMaxima.length >= MAX_ALTERNATIVES) {
        invalid('portable pattern has too many alternatives')
      }
    }
    return Math.max(...branchMaxima)
  }

  private scanBranch(): number {
    let atomCount = 0
    let quantifiedCount = 0
    let maximumLength = 0
    let previousQuantified = false
    let previousCharacters: AcceptedCharacters | null = null

    while (
      this.index < this.characters.length &&
      this.characters[this.index] !== '|'
    ) {
      const characters = this.scanAtom()
      atomCount += 1
      if (atomCount > MAX_ATOMS_PER_ALTERNATIVE) {
        invalid('portable pattern branch has too many atoms')
      }

      const [maximumRepetition, quantified] = this.scanQuantifier()
      if (quantified) {
        quantifiedCount += 1
        if (quantifiedCount > MAX_QUANTIFIED_ATOMS_PER_ALTERNATIVE) {
          invalid('portable pattern branch has too many quantified atoms')
        }
        if (
          previousQuantified &&
          previousCharacters !== null &&
          overlaps(previousCharacters, characters)
        ) {
          invalid('adjacent quantified atoms accept overlapping characters')
        }
      }

      maximumLength += maximumRepetition
      if (maximumLength > MAX_BRANCH_MATCH_LENGTH) {
        invalid('portable pattern branch maximum length is too large')
      }
      previousQuantified = quantified
      previousCharacters = characters
    }

    if (atomCount === 0) invalid('portable pattern alternatives cannot be empty')
    return maximumLength
  }

  private scanAtom(): AcceptedCharacters {
    const character = this.characters[this.index]
    if (character === undefined) invalid('portable pattern atom is missing')
    if (character === '.') {
      this.index += 1
      return ANY_CHARACTER
    }
    if (character === '[') return this.scanCharacterClass()
    if (character === '\\') {
      const literal = this.scanEscape()
      const codePoint = requireCodePoint(literal)
      return { intervals: [[codePoint, codePoint]], matchesAny: false }
    }
    if (FORBIDDEN_UNESCAPED.has(character)) {
      invalid(`unescaped metacharacter ${JSON.stringify(character)} is not portable`)
    }

    this.index += 1
    const codePoint = requireCodePoint(character)
    return { intervals: [[codePoint, codePoint]], matchesAny: false }
  }

  private scanEscape(): string {
    this.index += 1
    if (this.index === this.characters.length) {
      invalid('portable pattern ends with an incomplete escape')
    }
    const escaped = this.characters[this.index]
    if (escaped === undefined || !ESCAPABLE_METACHARACTERS.has(escaped)) {
      invalid('only literal metacharacters may be escaped')
    }
    this.index += 1
    return escaped
  }

  private scanCharacterClass(): AcceptedCharacters {
    this.index += 1
    const items: Array<readonly [string, boolean]> = []
    while (
      this.index < this.characters.length &&
      this.characters[this.index] !== ']'
    ) {
      const character = this.characters[this.index]
      if (character === '\\') {
        items.push([this.scanEscape(), true])
        continue
      }
      if (character === '[' || (character === '^' && items.length === 0)) {
        invalid('nested or negated character classes are not portable')
      }
      if (character === undefined) invalid('portable character class item is missing')
      items.push([character, false])
      this.index += 1
    }

    if (this.index === this.characters.length) {
      invalid('portable character class is not closed')
    }
    this.index += 1
    if (items.length === 0) invalid('portable character classes cannot be empty')

    const intervals: Array<readonly [number, number]> = []
    let itemIndex = 0
    while (itemIndex < items.length) {
      const item = items[itemIndex]
      if (item === undefined) invalid('portable character class item is missing')
      const [character, escaped] = item
      const separator = items[itemIndex + 1]
      const rangeEnd = items[itemIndex + 2]
      if (
        rangeEnd !== undefined &&
        separator?.[0] === '-' &&
        separator[1] === false &&
        character !== '-'
      ) {
        const end = rangeEnd[0]
        const startPoint = requireCodePoint(character)
        const endPoint = requireCodePoint(end)
        if (end === '-' || startPoint > endPoint) {
          invalid('portable character class contains a descending range')
        }
        intervals.push([startPoint, endPoint])
        itemIndex += 3
        continue
      }
      if (
        character === '-' &&
        !escaped &&
        itemIndex !== 0 &&
        itemIndex !== items.length - 1
      ) {
        invalid('portable character class contains a malformed range')
      }
      const codePoint = requireCodePoint(character)
      intervals.push([codePoint, codePoint])
      itemIndex += 1
    }
    return { intervals, matchesAny: false }
  }

  private scanQuantifier(): readonly [number, boolean] {
    if (this.index === this.characters.length) return [1, false]
    if (this.characters[this.index] === '?') {
      this.index += 1
      return [1, true]
    }
    if (this.characters[this.index] !== '{') return [1, false]

    const closing = this.characters.indexOf('}', this.index + 1)
    if (closing === -1) invalid('portable repetition is not closed')
    const body = this.characters.slice(this.index + 1, closing).join('')
    const parts = body.split(',')
    let lower: number
    let upper: number
    if (parts.length === 1 && isAsciiDigits(parts[0] ?? '')) {
      lower = parseBound(parts[0] ?? '')
      upper = lower
    } else if (
      parts.length === 2 &&
      isAsciiDigits(parts[0] ?? '') &&
      isAsciiDigits(parts[1] ?? '')
    ) {
      lower = parseBound(parts[0] ?? '')
      upper = parseBound(parts[1] ?? '')
    } else {
      invalid('portable repetition must be {m} or {m,n}')
    }
    if (lower > upper) invalid('portable repetition bounds are descending')
    if (upper > MAX_REPETITION) invalid('portable repetition upper bound is too large')
    this.index = closing + 1
    return [upper, true]
  }
}

export function compilePortablePattern(source: string): PortablePattern {
  const sourceCharacters = [...source]
  if (sourceCharacters.length === 0) invalid('portable pattern cannot be empty')
  if (sourceCharacters.length > MAX_PATTERN_LENGTH) {
    invalid('portable pattern source is too long')
  }

  const maxMatchLength = new PatternScanner(source).scan()
  let compiled: RegExp
  try {
    // Unicode mode makes one dot/class atom consume one code point, matching Python. JavaScript's
    // Unicode grammar rejects an escaped hyphen outside classes, so normalize only that already
    // validated literal spelling before compilation.
    compiled = new RegExp(`^(?:${unicodeRegexSource(source)})$`, 'u')
  } catch {
    invalid('portable pattern could not be compiled')
  }

  return Object.freeze({
    source,
    maxMatchLength,
    matches(value: string): boolean {
      if ([...value].length > maxMatchLength) return false
      const match = compiled.exec(value)
      return match !== null && match[0].length === value.length
    },
  })
}

function overlaps(left: AcceptedCharacters, right: AcceptedCharacters): boolean {
  if (left.matchesAny || right.matchesAny) return true
  return left.intervals.some(([leftStart, leftEnd]) =>
    right.intervals.some(
      ([rightStart, rightEnd]) => leftStart <= rightEnd && rightStart <= leftEnd,
    ),
  )
}

function unicodeRegexSource(source: string): string {
  const characters = [...source]
  let inClass = false
  let normalized = ''
  for (let index = 0; index < characters.length; index += 1) {
    const character = characters[index]
    if (character === '\\') {
      const escaped = characters[index + 1]
      if (escaped === undefined) return source
      normalized += escaped === '-' && !inClass ? String.raw`\x2D` : `\\${escaped}`
      index += 1
      continue
    }
    if (character === '[') inClass = true
    if (character === ']') inClass = false
    // Python dot excludes LF only; JavaScript dot additionally excludes CR/U+2028/U+2029.
    normalized += character === '.' && !inClass ? String.raw`[^\n]` : character
  }
  return normalized
}

function isAsciiDigits(value: string): boolean {
  return value.length > 0 && [...value].every((character) => character >= '0' && character <= '9')
}

function parseBound(value: string): number {
  let result = 0
  for (const character of value) {
    result = result * 10 + (character.codePointAt(0) ?? 48) - 48
    if (result > MAX_REPETITION) return result
  }
  return result
}

function requireCodePoint(character: string): number {
  const codePoint = character.codePointAt(0)
  if (codePoint === undefined) invalid('portable pattern character is missing')
  return codePoint
}

function invalid(diagnostic: string): never {
  throw new PortablePatternError(diagnostic)
}
